Ensures DAG params order regardless of backend

Usiel opened the pull request:
Fixes https://github.com/apache/airflow/issues/40154

This change adds an extra attribute to the serialized DAG param objects which helps us decide the order of the deserialized params dictionary later even if the backend messes with us.

I decided not to limit this just to MySQL since the operation is inexpensive and may turn out to be helpful (edit: Postgres is also affected).

### Tests

I made sure the new test fails with the old implementation + MySQL/Postgres. I assume this test will be executed with MySQL somewhere in the build actions?

Usiel committed:
Ensures DAG params order regardless of backend

Fixes https://github.com/apache/airflow/issues/40154

This change adds an extra attribute to the serialized DAG param objects which helps us decide
the order of the deserialized params dictionary later even if the backend messes with us.

I decided not to limit this just to MySQL since the operation is inexpensive and may turn
out to be helpful.

I made sure the new test fails with the old implementation + MySQL. I assume this test will be
executed with MySQL somewhere in the build actions?

jedcunningham commented on airflow/serialization/serialized_objects.py:
```diff
@@ -831,14 +831,17 @@ def is_serialized(val):
         serialized_params = {}
-        for k, v in params.items():
+        for idx, item in enumerate(params.items()):
+            k, v = item
```
```suggestion
        for idx, (k, v) in enumerate(params.items()):
```

jedcunningham commented on tests/models/test_serialized_dag.py:
```diff
@@ -206,6 +206,23 @@ def test_get_dag_dependencies_default_to_empty(self, dag_dependencies_fields):
 
+    def test_order_of_dag_params_is_stable(self):
+        """
+        https://github.com/apache/airflow/issues/40154
```
```suggestion
```

We can always find the why with git blame. I don't think this case is noteworthy enough to warrant a permanent link here.

Usiel committed:
Removes GitHub reference

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

Usiel committed:
Serialize DAG params as array of tuples to ensure ordering

Alternative to previous approach: We serialize the DAG params dict as a list of tuples which _should_ keep their ordering regardless of backend.

Backwards compatibility is ensured because if `encoded_params` is a `dict` (not the expected `list`) then `dict(encoded_params)` still works.

uranusjr commented on airflow/serialization/serialized_objects.py:
```diff
@@ -847,10 +847,10 @@ def _serialize_params_dict(cls, params: ParamsDict | dict):
         """Deserialize a DAG's Params dict."""
         op_params = {}
-        for k, v in encoded_params.items():
+        for k, v in dict(encoded_params).items():
```
```suggestion
        for k, v in encoded_params:
```

uranusjr replied:
Ah or is this so we can handle both the old and new cases? Converting to dict seems unnecessary to me though. Maybe

```python
if isinstance(encoded_params, collections.abc.Mapping):
    encoded_param_pairs = encoded_params.items()
else:
    encoded_param_pairs = encoded_params
for k, v in encoded_param_pairs:
    ...
```

Usiel replied:
> Ah or is this so we can handle both the old and new cases? 

Yeah, exactly. Yep I'll add that bit of code, makes more clear what we're doing here.

uranusjr commented on airflow/serialization/serialized_objects.py:
```diff
@@ -847,10 +847,10 @@ def _serialize_params_dict(cls, params: ParamsDict | dict):
 
     @classmethod
-    def _deserialize_params_dict(cls, encoded_params: dict) -> ParamsDict:
+    def _deserialize_params_dict(cls, encoded_params: list) -> ParamsDict:
```
Should change this to `list[tuple[str, Any]]` I guess?

Usiel committed:
Make backwards compatibility more explicit

Based on suggestions by @uranusjr with an additional fix to make mypy happy.
